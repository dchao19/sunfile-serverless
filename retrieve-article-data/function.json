{
    "disabled": false,
    "scriptFile": "retrieve-article-data.js",
    "bindings": [
        {
            "name": "respond to /api/article/data",
            "type": "httpTrigger",
            "direction": "in",
            "route": "article/data",
            "authLevel": "anonymous",
            "methods": [
                "post"
            ]
        },
        {
            "name": "create http response",
            "type": "http",
            "direction": "out"
        }
    ]
}