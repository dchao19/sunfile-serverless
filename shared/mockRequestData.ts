import { IHttpRequest, HttpMethod } from "./IContext";

export function mockRequestData(
    originalUrl: string,
    method: HttpMethod,
    query: { [s: string]: string },
    body: { [s: string]: any },
    headers: { [s: string]: any },
    rawBody: string
): IHttpRequest {
    return {
        originalUrl,
        method,
        query,
        headers,
        body,
        rawBody: ""
    };
}
