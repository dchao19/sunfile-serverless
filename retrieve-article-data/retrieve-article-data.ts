import { IHttpContext } from "../shared/IContext";
import axios from "axios";
import * as jQuery from "jquery";
import { JSDOM } from "jsdom";

export async function run(context: IHttpContext) {
    context.log("retrieve-article-data function processed a request");
    const { window } = new JSDOM("");

    var $ = require("jquery")(window) as JQueryStatic;
    var preStrip = $("<div/>").html(context.req.rawBody);
    var html = preStrip
        .find(".last-update,.fyre,style,script,.layout-detail-page__footer,.teaser__byline")
        .remove()
        .end()
        .html();

    let watsonDataRequest = axios.post(
        "https://gateway.watsonplatform.net/natural-language-understanding/api/v1/analyze?version=2017-02-27",
        {
            html,
            language: "en",
            return_analyzed_text: true,
            features: {
                metadata: {},
                keywords: {
                    limit: 5
                }
            }
        },
        {
            headers: {
                "Content-Type": "application/json"
            },
            auth: {
                username: process.env.WATSON_USERNAME,
                password: process.env.WATSON_PASSWORD
            }
        }
    );

    try {
        let watson = await watsonDataRequest;

        const paragraphs = watson.data.analyzed_text.split("\n").map(paragraphContent => {
            return {
                paragraphContent
            };
        });

        const keywords = watson.data.keywords.map(keyword => keyword.text);

        context.res = {
            message: "success",
            success: true,
            result: {
                title: watson.data.metadata.title,
                keywords,
                text: watson.data.analyzed_text,
                paragraphs,
                pubDate: watson.data.metadata.publication_date
            }
        };
    } catch (error) {
        context.res = {
            success: false,
            errorMessage: "An internal server error ocurred while parsing the article",
            errorCode: 500
        };
    }
}
