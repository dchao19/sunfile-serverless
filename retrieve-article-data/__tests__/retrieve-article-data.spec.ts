import { run } from "../retrieve-article-data";
import { mockRequestData } from "../../shared/mockRequestData";
import { IHttpContext } from "../../shared/IContext";
import axios from "axios";

describe("retrieve-article-data tests", async () => {
    test("should return the correct successful response", async () => {
        const response = await axios.post("http://localhost:7071/api/article/data");
        expect(response.data).toMatchSnapshot();
    });
});
