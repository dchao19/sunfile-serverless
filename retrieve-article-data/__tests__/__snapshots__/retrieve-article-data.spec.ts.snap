// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`retrieve-article-data tests should return the correct successful response 1`] = `
Object {
  "errorCode": 500,
  "errorMessage": "An internal server error ocurred while parsing the article",
  "success": false,
}
`;
